apps/src/music/MiniPlayerView.tsx

```diff
@@ -0,0 +1,109 @@
+import React, {useCallback, useEffect, useRef, useState} from 'react';
+import {Channel} from '../lab2/types';
+import MusicPlayer from './player/MusicPlayer';
+import MusicBlocklyWorkspace from './blockly/MusicBlocklyWorkspace';
+import Simple2Sequencer from './player/sequencer/Simple2Sequencer';
+import {RemoteSourcesStore, SourcesStore} from '../lab2/projects/SourcesStore';
+import {loadLibrary} from './utils/Loader';
+import MusicLibrary from './player/MusicLibrary';
+import {setUpBlocklyForMusicLab} from './blockly/setup';
+import Lab2Registry from '../lab2/Lab2Registry';
+
+interface MiniPlayerViewProps {
+  projects: Channel[];
+  libraryName: string;
+}
+
+const MiniPlayerView: React.FunctionComponent<MiniPlayerViewProps> = ({
+  projects,
+  libraryName,
+}) => {
+  const playerRef = useRef<MusicPlayer>(new MusicPlayer());
+  const workspaceRef = useRef<MusicBlocklyWorkspace>(
+    new MusicBlocklyWorkspace()
+  );
+  const sequencerRef = useRef<Simple2Sequencer>(new Simple2Sequencer());
+  const sourcesStoreRef = useRef<SourcesStore>(new RemoteSourcesStore());
+  const [isLoading, setIsLoading] = useState(true);
+
+  // Setup library and workspace on mount
+  const onMount = useCallback(async () => {
+    setUpBlocklyForMusicLab();
+    workspaceRef.current.initHeadless();
+    const library = await loadLibrary(libraryName);
+    MusicLibrary.setCurrent(library);
+    setIsLoading(false);
+  }, [libraryName]);
+
+  useEffect(() => {
+    onMount();
+  }, [onMount]);
+
+  // This is the main function that is called when a song is played in the mini player
+  // Loads code from the server, compiles the song, executes it to generate events,
+  // and then plays the events.
+  // Optimization: cache code and/or compiled song after played once.
```

Comment:
Is this a potential optimization, or one already in place?

Reply:
Oh this is a potential optimization, left a note here to add this as a follow-up. Currently we would always fetch sources for a project even if it's already been played.